test/test_runner.go

```diff
@@ -394,12 +417,30 @@ func recoverPanics(onError func(error)) {
 
 func (r *TestRunner) parseCheckAndInterpret(script string) (*interpreter.Program, *interpreter.Interpreter, error) {
 	config := runtime.Config{
-		CoverageReport: r.coverageReport,
+		AccountLinkingEnabled:        true,
+		AttachmentsEnabled:           true,
+		CapabilityControllersEnabled: true,
+		CoverageReport:               r.coverageReport,
 	}
 	env := runtime.NewBaseInterpreterEnvironment(config)
+	r.testRuntime = runtime.NewInterpreterRuntime(config)
+
+	r.testFramework = NewTestFrameworkProvider(
+		r.fileResolver,
+		env,
+		r.coverageReport,
+		r.testRuntime,
+	)
+	backend, ok := r.testFramework.NewEmulatorBackend().(*EmulatorBackend)
+	if !ok {
+		panic(fmt.Errorf("failed to retrieve EmulatorBackend"))
+	}
+	backend.fileResolver = r.fileResolver
+	backend.contracts = r.contracts
+	r.backend = backend
```

Comment:
The fact that
```go
env := runtime.NewBaseInterpreterEnvironment(config)
```
returns a non-exported type (`*runtime.interpreterEnvironment`), makes this somewhat difficult.
Pretty much everything has to be configured in `parseCheckAndInterpret`, since only in this method we can use the `env` variable and configure it properly, e.g. with:
```go
env.CheckerConfig.ImportHandler
env.CheckerConfig.ContractValueHandler
env.InterpreterConfig.ImportLocationHandler
env.InterpreterConfig.ContractValueHandler
env.InterpreterConfig.Storage = r.storage
...
```

We also utilize many option methods, e.g.
```go
WithImportResolver
WithFileResolver
WithCoverageReport
WithRandomSeed
WithContracts
```
So I'm not sure how safe it is to initialize everything in `NewTestRunner`. I agree though that the `parseCheckAndInterpret` probably needs a refactor.

Reply:
Not quite sure I follow, sorry.
Couldn't this code be moved as-is to the constructor?
i.e: As I understood, none of these codes relies on the parameter (`script`) of this function, so this 'configuring' of the environment can be a one-time operation per test-runner.

Reply:
The ultimate goal of the function is to parse / check and interpret the given `script` parameter.
```go
program, err := r.testRuntime.ParseAndCheckProgram([]byte(script), ctx)
if err != nil {
	return nil, nil, err
}
```
To call `ParseAndCheckProgram`, we first need a `runtime.Context`:
```go
ctx := runtime.Context{
	Interface:   r.backend.blockchain.NewScriptEnvironment(),
	Location:    testScriptLocation,
	Environment: env,
}
```
To instantiate that `ctx`, we also need to create the `backend`, `testFramework` and `env`:
```go
r.testFramework = NewTestFrameworkProvider(
	r.fileResolver,
	env,
	r.coverageReport,
)
backend, ok := r.testFramework.EmulatorBackend().(*EmulatorBackend)
```

After `ParseAndCheckProgram` and before `env.Interpret`, we also need to clear the storage:
```go
// Set the storage after checking, because `ParseAndCheckProgram` clears the storage.
r.storage = runtime.NewStorage(ctx.Interface, nil)
env.InterpreterConfig.Storage = r.storage
```

If I were to add a field `testEnv` of type `runtime.Environment`, I would no longer be able to access neither `CheckerConfig` nor `InterpreterConfig`. `*runtime.interpreterEnvironment` is a non-exported type, so I can't use that type in the field type of a struct.
Sorry for the lengthy reply, I hope this clarifies the obstacle here :sweat_smile:

Reply:
oh so the issue is storage getting reset. I'm not sure why the `ParseAndCheckProgram` method does that.

But I managed to bypass that and directly use the `ParseAndCheckProgram` method of `Environment`. Did quick refactor to test it: https://github.com/onflow/cadence-tools/commit/3ac83b4f973b11e863b208eeef2acb33beb77d32

Realized it still needed some more refactoring to completely move this to the initializer. For e.g: the coverage report is set via a setter method. We can leave that part for now.

Reply:
Updated in https://github.com/onflow/cadence-tools/pull/210/commits/5cf8589e143d16de7e7b2dd97d82891985b27e3d